### PR TITLE
Add Scaleway provider to linuxkit

### LINKED ISSUE #2487: Add scaleway support
Scaleway currently supports launching instances using docker images, so I think there shouldn’t be much needed to make this work.

### PR DESCRIPTION
**Pull request by Sh4d1**
Fixes #2487

**- What I did**
I added support for scaleway provider. Right now I only did the push command. 
Some modifications are still needed, but I wanted some feedback.
**- How I did it**
I used to Scaleway Go API to implement the provider.
The tricky part was the process of the image creation since Scaleway relies on snapshot to make images. 
Basically I did:
  * Create a temporary instance with a second volume attached
  * Copy the iso file to the instance via ssh
  * Write the iso to the second volume using dd command
  * Terminate the instance
  * Snapshot the second volume
  * Create the image from snapshot

**- How to verify it**
Pull and try :grin: 

**- Description for the changelog**

Add scaleway support for LinuxKit

**- A picture of a cute animal (not mandatory but encouraged)**
cute cat

### CODE REVIEW

File: src/cmd/linuxkit/push_scaleway.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func pushScaleway(args []string) {
+	flags := flag.NewFlagSet("gcp", flag.ExitOnError)
```
Comment (rn) [inline src/cmd/linuxkit/push_scaleway.go]: this looks wrong

**Comment by rn** [comment]
I'm easy either way. I don't know scaleway at all so have no strong opinion on how one might implement it.

Just for people to have a look at the code maybe just add a link to your clone to the tracking issue?

**Comment by Sh4d1** [comment]
Yep will do! :)

**Comment by Sh4d1** [comment]
@rn The CI is failing due to lack of support of creack/goselect for s390x. Should it be implemented in the package in order to merge the PR? Or is it okay if it does not compile for s390x. I'm guessing the first one but not sure.

**Comment by rn** [comment]
The linuxkit binary should compile for all platforms. Maybe make the scaleway support conditional for arm and x86?

**Comment by Sh4d1** [comment]
@rn Quick PR on the repo did the trick :smile: 
Also should there be tests for providers? I saw one for GCP but seems empty..

**Comment by rn** [comment]
we currently don't have tests for any of the providers except for GCP. It would be good to do at least a build test of the scaleway example, though, if you haven't done that already (haven't looked at the PR in detail).

**Comment by Sh4d1** [comment]
@rn Yep I see. I'll look into the tests during the week I guess. Regarding the PR it is almost ready. Just waiting for a PR to be merged in moul/gotty-client to fix a terminal display issue after exiting the serial output. But it is working if you want to test :)

File: pkg/metadata/provider_scaleway.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+// Probe checks if we are running on Scaleway
+func (p *ProviderScaleway) Probe() bool {
+	// Getting the conf should always work...
+	_, err := scalewayGet(scalewayMetadataURL + "conf")
+	if err != nil {
+		log.Printf(err.Error())
+		return false
+	}
+
+	// we are on Scaleway so we need to send a request to tell that the instance has correclty booted
```
Comment (justincormack) [inline pkg/metadata/provider_scaleway.go]: typo "correclty"!

File: pkg/metadata/provider_scaleway.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to write private_ip: %s", err)
+	}
+
+	if err := p.handleSSH(metadata); err != nil {
+		log.Printf("Scaleway: Failed to get ssh data: %s", err)
+	}
+
+	// Generic userdata
+	// should use a port below 1024, TODO fix that, getting forbidden for now
```
Comment (justincormack) [inline pkg/metadata/provider_scaleway.go]: port below 1024 for what?

File: pkg/metadata/provider_scaleway.go (hunk 2 above)
Comment (Sh4d1) [reply pkg/metadata/provider_scaleway.go]: Oops, I'll fix that!

File: pkg/metadata/provider_scaleway.go (hunk 3 above)
Comment (Sh4d1) [reply pkg/metadata/provider_scaleway.go]: To get scaleway's userdata the source port must be below 1024 (to ensure that the calling has root privileges I guess)

File: src/cmd/linuxkit/vendor.conf [hunk 4]
```diff
@@ -53,3 +53,11 @@ golang.org/x/text c4d099d611ac3ded35360abf03581e13d91c828f
...
 google.golang.org/grpc 0713829b980f4ddd276689a36235c5fcc82a21bf
 gopkg.in/yaml.v2 a3f3340b5840cee44f372bddb5880fcbc419b46a
+github.com/scaleway/go-scaleway d2f0e78b23de36d1ae719dd9919cc53da8ebb928
+github.com/moul/anonuuid c6987e46f8a0231504bcd402962749f8dd9970b0
+golang.org/x/sync 1d60e4601c6fd243af51cc01ddf169918a5407ca
+github.com/renstrom/fuzzysearch 7a8f9a1c4bed53899ecd512daeaf8207cc454156
+github.com/moul/gotty-client c1b82f1af3ebb2febba268415cb1c90686078f80
+github.com/gorilla/websocket 21ab95fa12b9bdd8fecf5fa3586aad941cc98785
+github.com/creack/goselect 58854f77ee8d858ce751b0a9bcc5533fef7bfa9e
+github.com/moby/moby d339130f307634a0edd6aef29b6934c3934e9ba2
```
Comment (justincormack) [inline src/cmd/linuxkit/vendor.conf]: whats the moby/moby dep for?

Comment (justincormack) [reply src/cmd/linuxkit/vendor.conf]: oh, pkg/term.. hmm might be good to update to use containerd/console

**Comment by justincormack** [comment]
Ah thats fixed, you could rebase the scaleway vendor...

**Comment by Sh4d1** [comment]
Yep that's fixed along with the typo. Regarding moul/gotty-client, should I try to push a PR with containerd/console instead of moby pkg/term?

**Comment by justincormack** [comment]
Generally I would recommend using the containerd packages, they are better maintained and more up to date...

**Comment by Sh4d1** [comment]
@justincormack okay so I requested change on gotty-client:  https://github.com/moul/gotty-client/pull/64 
Once merged (I hope :smile: ), will they still be some changes to be made?

File: pkg/metadata/provider_scaleway.go (hunk 3 above)
Comment (Sh4d1) [reply pkg/metadata/provider_scaleway.go]: One solution could be to call `curl` from the code. Otherwise I guess it is possible to do it purely in Go but I searched and found nothing..

Comment (justincormack) [reply pkg/metadata/provider_scaleway.go]: oh, just add `CAP_NET_BIND_SERVICE` to the metadata package capabilities in here https://github.com/linuxkit/linuxkit/blob/master/pkg/metadata/build.yml#L13

Comment (Sh4d1) [reply pkg/metadata/provider_scaleway.go]: Yep but the main question remains. Should I prefer a pure go implemention, or is it okay to call `curl` from the go code?

Comment (justincormack) [reply pkg/metadata/provider_scaleway.go]: Go please.

Comment (justincormack) [reply pkg/metadata/provider_scaleway.go]: its pretty simple eg https://gist.github.com/higebu/96c09c8f03113997fd9a5338ca99764a but you need to change the capabilities or it will fail

File: pkg/metadata/provider_scaleway.go [hunk 5]
```diff
@@ -183,6 +186,47 @@ func scalewayGet(url string) ([]byte, error) {
 	return body, nil
 }
 
+// scalewayGetUserdata returns the userdata of the server, differs from scalewayGet since the source port has to be below 1024 in order to work
+func scalewayGetUserdata() ([]byte, error) {
```
Comment (Sh4d1) [inline pkg/metadata/provider_scaleway.go]: @justincormack seems good for you?

**Comment by Sh4d1** [comment]
@justincormack just bumped the dependencies, no more moby/moby in the code :smile:

**Comment by justincormack** [comment]
Can you squash the commits a bit (or completely), 17 commits seems a bit excessive!

**Comment by Sh4d1** [comment]
@justincormack done!

**Comment by Sh4d1** [comment]
@rn done :smile:

**Comment by rn** [comment]
Thanks for the quick turnaround. I've build and pushed the updated metadata package. Could you add a commit which updates the YAML files to the new version. There is a script, so in the top-level directory something like:
```
./scripts/update-component-sha.sh --image linuxkit/metadata:aa4f93f6695ebe21a3539b5effed27889d7c4dfb
```
should do the right thing.

Once that update PR passes CI, I'm happy to merge it.

**Comment by Sh4d1** [comment]
@rn updating all examples files using metadata package creates conflicts. Should I update only `scaleway.yml`?

**Comment by rn** [comment]
argh, we had the same dance earlier this week on a different PR.Sorry, this is going to be a little bit tedious. Could you:

- Drop commit with changes to the YAML files.
- Rebase your PR to master
- Force push

I will then rebuild the meta data package and your can do another update to the YAMLs with a force push. I'm surprised git can't merge the changes...

**Comment by Sh4d1** [comment]
@rn yes, it's done!

**Comment by rn** [comment]
Thanks. The hash actually hasn't change, but the meta-data (git commit etc) on the package should now be correct. Can push an update to the YAML files again (same script invocation as above).
Thanks

**Comment by rn** [comment]
actually, different hash, sorry:
```
./scripts/update-component-sha.sh --image linuxkit/metadata:ae402c118338c0285625c0094525eb0a90085a58
```

**Comment by Sh4d1** [comment]
@rn perfect, updated with no conflict :tada:
